Expand C install guide (#5327)
Qiskit/documentation: docs/guides/install-c-api.mdx

This section provides build instructions for Windows systems.

There are two independent ways to use the C API on Windows:

* **Build Python extension modules** that use the Qiskit C API. Follow Steps 1-5. This path uses the C headers shipped with the `qiskit` Python package and does not require Rust or cbindgen.

> * **Build the standalone C library** to link against from a pure C program, as in the UNIX-like section. Complete Step 1, then skip to [Build the standalone library](#build-the-standalone-library), which lists its additional requirements.

Reviewer: What does UNIX-like section refer to? (Also on line 570)

Author: The first section in the topic. I'll add a link.

</AccordionItem>
<AccordionItem title="VS Code">
Download VS Code from the [Visual Studio Code](https://code.visualstudio.com/) site, or run `winget install Microsoft.VisualStudio.Code`. Run the downloaded executable file to install VS Code.

After it is installed, continue with these steps:
1. Open VS Code
2. Click File -> Open Folder, then select your workspace (for example, `C:\workspace`).
3. Click Terminal -> New terminal to open a PowerShell terminal.
4. Click the Extensions icon on the left or press Ctrl+Shift+X. In the Extensions window, search for and install `ms-python.python`, `ms-toolsai.jupyter`, and `ms-vscode.cpptools`.

Run the rest of the commands in this guide in the VS Code terminal, unless otherwise instructed. The VS Code terminal defaults to PowerShell, preventing mix-ups with the Windows built-in command window. 

Set the workspace variable.  For example, if your workspace is called `workspace`, run the following:

```powershell
$WORKSPACE = "C:\workspace"   # change to your workspace path
mkdir $WORKSPACE -Force
cd $WORKSPACE
```
</AccordionItem>
<AccordionItem title="Python 3.12 (recommended version)">

> Python 3.12 is recommended because it has the best wheel availability for qiskit-aer and other dependencies. 3.10 and 3.11 also work, but 3.13 or later might lack pre-built wheels for some packages.

Reviewer: ```suggestion
Python 3.12 is recommended because it has the best wheel availability for `qiskit-aer` and other dependencies. 3.10 and 3.11 also work, but 3.13 or later might lack pre-built wheels for some packages.
```

    }
} catch {}
if (-not $PYTHON_EXE) {
    try {
        $ver  = (python --version 2>&1) -replace "Python ", ""
        $bits = python -c "import platform; print(platform.architecture()[0])"
        $path = python -c "import sys; print(sys.executable)"
        if ($path -match "(?i)(anaconda|miniconda|miniforge|mambaforge|[/\\]conda[/\\]|[/\\]envs[/\\])") {
            Write-Host "Skipping conda-managed Python at: $path"
        } elseif ([version]$ver -ge [version]"3.10" -and $bits -eq "64bit") {
            $PYTHON_EXE = $path
            Write-Host "Python $ver (64-bit) found: $PYTHON_EXE"
        } else {
            Write-Host "Skipping ($ver, $bits) — need 3.10+ 64-bit"
        }
    } catch {}
}

# Install Python 3.12 if it wasn't found.
if (-not $PYTHON_EXE) {
    Write-Host "Not found. Installing Python 3.12..."
    winget install Python.Python.3.12
    Write-Host "Close and reopen the terminal, then rerun this snippet."
}
if ($PYTHON_EXE -and ($PYTHON_EXE -match '[^\x20-\x7E]')) {
    Write-Warning "Python path has non-ASCII characters. Keep your workspace on an ASCII path."
}
if ($PYTHON_EXE) { Write-Host "`$PYTHON_EXE = '$PYTHON_EXE'" }
```

> If the download didn't work, or you prefer manual download, comment out the code lines that install Python 3.12, then download Python 3.12 from the [Python](https://www.python.org/downloads/).  Run the executable to install Python.  Select "Add Python to PATH" during the install, then rerun the above snippet to make sure it's being found.

Reviewer: ```suggestion
If the download didn't work, or you prefer manual download, comment out the code lines that install Python 3.12, then download Python 3.12 from the [Python](https://www.python.org/downloads/) website.  Run the executable to install Python.  Select "Add Python to PATH" during the install, then rerun the above snippet to make sure it's being found.
```

<Admonition type="note" title="Notes">
- Do NOT use the Microsoft Store Python because it lacks C headers. If `python` opens the Microsoft Store, disable the alias by going to Windows Settings -> Apps -> Advanced app settings -> App execution aliases.

- Anaconda users: run `conda deactivate` until the (base) prefix disappears. If it won't go away, open a new terminal in VS Code.
</Admonition> 
</AccordionItem>

<AccordionItem title="Git (optional)">
    Only needed if you clone lab repositories.  Run `winget install Git.Git`.
</AccordionItem>

</Accordion>

### Step 2 - Set up the Python virtual environment with Qiskit

<Accordion>
<AccordionItem title="Reset the VSCode terminal">

> Open the VSCode terminal and reset it:

Reviewer: ```suggestion
<AccordionItem title="Reset the VS Code terminal">
Open the VS Code terminal and reset it:
```

    else { Write-Host "Python not found — complete Step 1.3 first." ; return }
    Write-Host "`$PYTHON_EXE = '$PYTHON_EXE'"
}
```
</AccordionItem>
<AccordionItem title="Create and activate a virtual environment">
Allow script execution (once per user), then create the virtual environment:

```powershell
Set-ExecutionPolicy -Scope CurrentUser RemoteSigned
cd $WORKSPACE
& $PYTHON_EXE -m venv .venv --prompt workspace
.\.venv\Scripts\Activate.ps1
```

If activation shows a red error about "scripts disabled", the `Set-ExecutionPolicy` line was not run.  Run it manually, then retry.

Your prompt should now show `(workspace)`. 

Verify:

```powershell
Get-Command python | Select-Object -First 1 -ExpandProperty Source
# -> workspace\.venv\Scripts\python.exe
```
</AccordionItem>
<AccordionItem title="Install packages">
```powershell
python -m pip install --upgrade pip setuptools wheel
pip install "qiskit[visualization]>=2.4.2"
pip install --prefer-binary qiskit-ibm-runtime qiskit-aer
pip install notebook ipykernel ipywidgets  # optional: run the following Python steps in Jupyter
```

> `--prefer-binary` avoids compiling qiskit-aer from source. If qiskit-aer still fails, try `pip install qiskit-aer --only-binary=:all:` or skip it.  Qiskit-aer is optional and is only needed for local simulation.

Reviewer: ```suggestion
`--prefer-binary` avoids compiling `qiskit-aer` from source. If `qiskit-aer` still fails, try `pip install qiskit-aer --only-binary=:all:` or skip it.  `qiskit-aer` is optional and is only needed for local simulation.
```

<Admonition type="note">
Do NOT run `pip install --upgrade qiskit` after setup. Upgrading to a new minor version breaks C extensions built against the old one.
</Admonition>
</AccordionItem>
</Accordion>

### Step 3 - Load the MSVC environment

Run the following code in each new Python session (for example, each time you restart a Jupyter kernel). It locates and loads the MSVC developer environment automatically so you don't need the x64 built-in command prompt.

> The example in this documentation builds C extension modules by using setuptools with MSVC. Generally, you define `QISKIT_PYTHON_EXTENSION`, include `qiskit.h`, and call `qk_import()` in your init function. Only headers from `qiskit.capi.get_include()` are needed at build time — no library is linked. See [Extend Qiskit in Python with C](/docs/guides/c-extension-for-python) for details.

Reviewer: Is "setuptools" a product/code element? Should we offset with `setuptools`?

```python
import qiskit.capi

inc = qiskit.capi.get_include().replace("\\", "/")
print(".vscode/c_cpp_properties.json — create this file with the content below:\n")
print('{')
print('    "version": 4,')
print('    "configurations": [')
print('        {')
print('            "name": "Win32",')
print(f'            "includePath": ["{inc}"],')
print('            "defines": ["QISKIT_PYTHON_EXTENSION"],')
print('            "compilerPath": "cl.exe",')
print('            "cStandard": "c11",')
print('            "intelliSenseMode": "windows-msvc-x64"')
print('        }')
print('    ]')
print('}')
```
</AccordionItem>
<AccordionItem title="Make the VS Code extensions file">
In VS Code, create `.vscode/extensions.json` with this content:

```json
{
    "recommendations": ["ms-python.python", "ms-toolsai.jupyter", "ms-vscode.cpptools"]
}
```
</AccordionItem>
</Accordion>
### Build the standalone library

This section builds the standalone C library, which is only needed if you want to compile and link pure C programs as in the UNIX-like section. In addition to the Step 1 prerequisites, this requires the following tools:

> * A Rust compiler: see for example the [guide on installing Qiskit from source](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#installing-qiskit-from-source)

Reviewer: Just a note that I was about to recommend pointing to https://quantum.cloud.ibm.com/docs/guides/install-qiskit-source instead, but it doesn't look like it goes into much detail for the Rust compiler part. We might want to consider beefing up the guide to include this? cc: @kaelynj

Update the [App Installer](https://apps.microsoft.com/detail/9nblggh4nns1) from the Microsoft Store, or use the relevant manual download links
</AccordionItem>
<AccordionItem title="`cl` is not recognized">
Rerun the MSVC load cell (Step 3), or use the x64 Native Tools Command Prompt
</AccordionItem>
<AccordionItem title="`python` opens Microsoft Store">
Go to Settings -> Apps -> Advanced app settings -> App execution aliases and turn off `python.exe`
</AccordionItem>
<AccordionItem title="`.ps1 cannot be loaded` / scripts disabled">
Run `Set-ExecutionPolicy -Scope CurrentUser RemoteSigned` then retry
</AccordionItem>
<AccordionItem title="`cannot open file 'qiskit.h'`">
Run `python -c "import qiskit.capi; print(qiskit.capi.get_include())"` and confirm the path exists
</AccordionItem>
<AccordionItem title="`qiskit.capi` not found">
Run `pip install "qiskit[visualization]~=2.4.2"`
</AccordionItem>
<AccordionItem title="`qiskit-aer` build fails">
Run `pip install qiskit-aer --only-binary=:all:`. If that also fails, use Python 3.12 or skip aer, which is optional
</AccordionItem>
<AccordionItem title="Build runs but import fails with a version error">

> Build and runtime Qiskit must share a minor version.  Reinstall with `pip install "qiskit~=2.4.2"`

Reviewer: What are "Build and runtime Qiskit"?

Author: Oooh - good catch.  Investigating.

Reviewer 2: The Qiskit you compile the extension against (build time) and the Qiskit you run it with (runtime) must be the same minor version. If you met the version error, reinstall the exact version you built against — check it with `pip show qiskit`, then `pip install "qiskit==<that-version>"`.

> Build and runtime Qiskit must share a minor version.  Reinstall with `pip install "qiskit~=2.4.2"`

Reviewer: ```suggestion
Be sure the the version for compiling (build time) and runtime version are the same minor version. Reinstall with `pip install "qiskit~=2.4.2"`
```

</AccordionItem>
<AccordionItem title="I rebuilt the C but the circuit did not change">
A C extension cannot be re-imported live. Restart the kernel, rerun the MSVC commands (Step 3), then rebuild
</AccordionItem>
<AccordionItem title="PowerShell script was blocked">
`Set-ExecutionPolicy -Scope CurrentUser RemoteSigned`
</AccordionItem>
<AccordionItem title="Path length errors">
Use a short root path on C:\ (for example, `C:\workspace`, or the path you set in the setup step), or enable Long Paths: Settings -> System -> For developers -> Long Paths
</AccordionItem>
<AccordionItem title="Conda is active (the prompt shows `(base)`) or builds behave oddly after using Anaconda">
Run `conda deactivate` until `CONDA_DEFAULT_ENV` and `CONDA_PREFIX` are both gone from the environment. Check with `$env:CONDA_PREFIX`. If they persist, open a fresh PowerShell (not an Anaconda prompt) and retry from Step 2.
</AccordionItem>
<AccordionItem title="The virtual environment was created from a conda-managed Python (check the `home =` line in `.venv\pyvenv.cfg`)">

> The virtual environment inherited conda's C runtime and cannot be fixed in place. Delete it and rebuild from a python.org Python. Run `Remove-Item -Recurse -Force .venv`, then rerun the Python detection snippet in Step 1 to set `$PYTHON_EXE`, then recreate the virtual environment (Step 2).

Reviewer: ```suggestion
The virtual environment inherited Conda's C runtime and cannot be fixed in-place. Delete it and rebuild from a python.org Python. Run `Remove-Item -Recurse -Force .venv`, then rerun the Python detection snippet in Step 1 to set `$PYTHON_EXE`, then recreate the virtual environment (Step 2).
```
What does "python.org Python" refer to - simply the latest Python from the website?

</AccordionItem>
<AccordionItem title="VS Build Tools cannot be installed (no admin rights)">
Administrator access is required. Ask your IT department for access
</AccordionItem>
<AccordionItem title="`WinError 32` / file in use during rebuild">
The .pyd is locked by the running kernel. Restart the kernel (Ctrl+Shift+P -> Jupyter: Restart Kernel), rerun Step 3, then rebuild
</AccordionItem>
<AccordionItem title="Commands silently do nothing (no error, no output)">
You might be in cmd.exe instead of PowerShell. Check your prompt: PowerShell shows `PS C:\>`, cmd shows `C:\>`. Open PowerShell from Start Menu or Win+X
</AccordionItem>
<AccordionItem title="MSVC install appears stuck">
The `--passive --wait` flags block PowerShell while the installer runs in the background. Check your taskbar for a "Visual Studio Installer" window. It can take 10-30 minutes to install
</AccordionItem>
</Accordion>
## Next steps

<Admonition type="tip" title="Recommendations">
  - Learn how to [extend Python with the Qiskit C API](/docs/guides/c-extension-for-python).
  - [Install Qiskit.](/docs/guides/install-qiskit)

>   - Explore the [Qiskit C API reference.](/docs/api/qiskit-c)

Reviewer: ```suggestion
  - [Install Qiskit](/docs/guides/install-qiskit).
  - Explore the [Qiskit C API reference](/docs/api/qiskit-c).
```
